docs/operators/retain.md

```diff
@@ -0,0 +1,256 @@
+| `on_error` | `send`           | The behavior of the operator if it encounters an error. See [on_error](/docs/types/on_error.md)                                                                                                                                          |
+| `if`       |                  | An [expression](/docs/types/expression.md) that, when set, will be evaluated to determine whether this operator should be used for the given entry. This allows you to do easy conditional parsing without branching logic with routers. |
+
+Example usage:
+<hr>
+retain fields in the body
+
+```yaml
+- type: retain
+    fields:
+      - key1
+      - key2
+```
+
+<table>
+<tr><td> Input Entry </td> <td> Output Entry </td></tr>
+<tr>
+<td>
+
+```json
+{
+  "resource": { },
+  "attributes": { },  
+  "body": {
+    "key1": "val1",
+    "key2": "val2",
+    "key3": "val3",
+    "key4": "val4"
+  }
+}
+```
+
+</td>
+<td>
+
+```json
+{
+  "resource": { },
+  "attributes": { },  
+  "body": {
+    "key1": "val1",
+    "key2": "val2"
+  }
+}
+```
+
+</td>
+</tr>
+</table>
+
+<hr>
+retain an object in the body
+
+```yaml
+- type: retain
+    fields:
+      - object
+```
+
+<table>
+<tr><td> Input record </td> <td> Output record </td></tr>
+<tr>
+<td>
+
+```json
+{
+  "resource": { },
+  "attributes": { },  
+  "body": {
+    "key1": "val1",
+    "object": {
+      "nestedkey": "val2",
+    }
+  }
+}
+```
+
+</td>
+<td>
+
+```json
+{
+  "resource": { },
+  "attributes": { },  
+  "body": {
+    "object": {
+      "nestedkey": "val2",
+    }
+  }
+}
+```
+
+</td>
+</tr>
+</table>
+
+<hr>
+retain fields from resource
+
+```yaml
+- type: retain
+    fields:
+      - key1
+      - key2
+```
+
+<table>
+<tr><td> Input record </td> <td> Output record </td></tr>
+<tr>
+<td>
+
+```json
+{
+  "resource": { 
+     "key1": "val1",
+     "key2": "val2",
+     "key3": "val3"
+  },
+  "attributes": { },  
+  "body": {
+    "key1": "val1",
+    }
+  }
+}
+```
+
+</td>
+<td>
+
+```json
+{
+  "resource": { 
+     "key1": "val1",
+     "key2": "val2",
+  },
+  "attributes": { },  
+  "body": { }
+  }
+}
+```
+
+</td>
+</tr>
+</table>
+
+<hr>
+retain fields from attributes
+
+```yaml
+- type: retain
+    fields:
+      - key1
+      - key2
+```
+
+<table>
+<tr><td> Input record </td> <td> Output record </td></tr>
+<tr>
+<td>
+
+```json
+{
+  "resource": { },
+  "attributes": { 
+     "key1": "val1",
+     "key2": "val2",
+     "key3": "val3"
+  },  
+  "body": { 
+    "key1": "val1",
+  }
+}
+```
+
+</td>
+<td>
+
+```json
+{
+  "resource": { },
+  "attributes": { 
+     "key1": "val1",
+     "key2": "val2",
+  },  
+  "body": { }
+}
+```
+
+</td>
+</tr>
+</table>
+
+<hr>
+retain fields from all sources
+
+```yaml
+- type: retain
+    fields:
+      - $resource.key1
+      - $attributes.key3
+      - key5
```

Comment:
I think that by supporting `resource` and `attributes`, we've actually introduced a bit of a design problem for this operator. 

Basically, I think it's generally unlikely that users will want to drop resource and attribute fields, so the primary use case is probably just for body.

I need to think on this one a bit.

Reply:
we could do something like this 
`
// Transform will apply the retain operation to an entry
func (p *RetainOperator) Transform(e *entry.Entry) error {
	newEntry := entry.New()
	containsRorA := false
	for _, field := range p.Fields {
		if reflect.TypeOf(field) == reflect.TypeOf(entry.BodyField{}) || reflect.TypeOf(field) == reflect.TypeOf(entry.ResourceField{}) {
			containsRorA = true
		}
	}
	if containsRorA {
		newEntry.Timestamp = e.Timestamp
		for _, field := range p.Fields {
			val, ok := e.Get(field)
			if !ok {
				return fmt.Errorf("retain: field does not exist: %s", field)
			}
			err := newEntry.Set(field, val)
			if err != nil {
				return err
			}
		}
		*e = *newEntry
		return nil
	}
	newEntry.Timestamp = e.Timestamp
	newEntry.Attributes = e.Attributes
	newEntry.Resource = e.Resource
	for _, field := range p.Fields {
		val, ok := e.Get(field)
		if !ok {
			return fmt.Errorf("retain: field does not exist: %s", field)
		}
		err := newEntry.Set(field, val)
		if err != nil {
			return err
		}
	}
	*e = *newEntry
	return nil
}
`
that way if they on't specify any attributes or resources we will just keep everything